[SYCL][ESIMD] Intel "Explicit SIMD" extension documentation draft.

kbobrovs opened the pull request:
This PR starts a series of PRs to integrate DPC++ "Explicit SIMD" extension for efficient Intel GPU programming. This PR provides overall description of the extenstion.
   Contributors (alphabetic order):
    Konst Bobrovsky
    Peter Caday
    Gang Chen
    Kai Yu Chen
    Ken Lueh
    Wei Pan
    Xinmin Tian

The upcoming changes will be spread across a number of components:
- Driver
`-fsycl-explicit-simd` option option added to tweak FE behavior slightly compared to normal SYCL. In the future it is expected that no such distinction will be necessary, and the need in extra -fsycl-explicit-simd option will go away.
- Front-End
    - recognize and mark explicit SIMD kernels and functions
    - allow non-constant globals used in device code
    - `__SYCL_EXPLICIT_SIMD__` macro is set by (device FE only)
    - LLVM passes are run (-disable-llvm-passes is not added) (device FE only)
- LLVM
    - A number of ESIMD-specific LLVM-IR passes are added
- SYCL RT - ESIMD-specific tweaks in accessors (using image buffers), 2D images,...
- New SIMD API (~9000 lines)
This is the largest piece implementing the essence of the ESIMD extension. It provides APIs to write device code in close to metal manner.
- llvm-spirv translator
- (will be external github project) GenXIntrinsics llvm project. Provides APIs for low-level intrinsic generation.

More description of changes in each component will come with respective PR.

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  static constexpr int length = __N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
```
Those should be defaulted. Why are these constexpr?

kychendev replied:
The first is defaulted. The others are implemented differently depending on it's device or host mode. It's marked constexpr to avoid dynamic initialization when arguments are all constant expressions in some tests.

mattkretz replied:
Having only `constexpr` ctors is not overly interesting. It becomes more interesting when all operations can be constant expressions as well. This allows to pre-compute certain values and be sure it doesn't cost at runtime. It's analogous to why you want constexpr `float` & `double` even though there were not usable as template arguments.
IIUC we will then also be able to use `simd` template arguments in C++20.

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &__Val);
+  constexpr simd(std::initializer_list<__Ty> __Ilist) noexcept;
+  constexpr simd(__Ty __Val, __Ty __Step = __Ty()) noexcept;
```
Why not take a generator expression like in `std::simd` instead? This is more flexible and obvious than this solution. 

```suggestion
  explicit simd(__Ty __Val) noexcept;
  template<class G> explicit simd(G&& gen) noexcept;
```

kychendev replied:
Currently this interface only supports a simple initilaizer pattern to generate a sequence based on the given initial value and incremental step, which is obvious enough. This can be mapped to efficient SIMD code with packed vector operand. For more flexible cases it's expected to pass in SIMD vector as argument or read it from memory, otherwise it may result in inefficient scalar code.

rolandschulz replied:
> which is obvious enough

What kind of argument is this? 

As explained above having an overload-set which both takes init-list and val+step is really error prone.

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
+  using type = simd<__Ty, __N>;
```
Why is this needed? It seems you can use the `simd` everywhere you are using this. Making it more readable. Also the current usage of this is inconsistent. You are using `simd` for the copy/move constructor but `type` for the assignment.

kychendev replied:
The type alias was declared for convenience. Will change it to just use simd.

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  constexpr type &operator=(type &&) &;
+
+  // Subscript operators.
+  __Ty operator[](int __i);
```
There is no point in having this non-const overload.

kychendev replied:
Will remove this from spec.

rolandschulz replied:
Actually why does this overload not return `simd_view`. This would allow you to use the subscript operator to write to an element.

kbobrovs replied:
added TODO to the headers

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  type operator--(int);
+
+  // Binary operators.
+  auto operator +(const type &__RHS) const;
```
No need to suddenly use auto.

kychendev replied:
For binary op the implementation will compute the intermediate result type, which may be different from the source operand data type.

rolandschulz replied:
Why would the return type ever be different for a binary operator?

kbobrovs replied:
added TODO to the headers

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  type &operator ^=(const type &__RHS);
+
+  // Compare operators.
+  simd<uint16_t, __N> operator >(const type &__RHS) const;
```
This return type is very weird. Why is there no `simd_mask` type? Why is the storage format of the mask not implementation defined?

kychendev replied:
In CM implementation simd mask type is mapped to vector of unsigned short.

rolandschulz replied:
But why expose that to a user?

kbobrovs replied:
added TODO to the headers

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  template <int Rep, int W> simd<__Ty, Rep * W> replicate(uint16_t __i);
+  template <int Rep, int VS, int W>
+  simd<__Ty, Rep * W> replicate(uint16_t __i);
+  template <int Rep, int VS, int W, int HS>
```
Template function with that many arguments are really ugly. Are you sure there isn't a better interface? And that users won't constantly forget what the correct order of the argument is? Some kind of templated builder pattern would be a bit more verbose but much more readable.

kbobrovs replied:
Can this new builder be added atop later? Can use this comment as a TODO for now.

rolandschulz replied:
Adding the other interface on-top only works if we want both interfaces. I personally don't. I wouldn't want to have to review code using this interface.

kbobrovs replied:
OK, can you suggest a change then to converge faster?

rolandschulz replied:
I suggest that replicate return an (implementation defined) replicate expression. And that allows to add the extra info (width, vstride, hstride). The replicate expression would have those 3 methods (each taking one template-arg) and the implicit conversion to the simd class. The simd class would only have the 1 replicate method taking the single `REP` template-arg and the `i` arg.
The user would use  (any of the extra method calls are optional)
```c++
s.replicate<R>(i).width<W>().vstride<VS>().hstride<HS>()
```

kbobrovs replied:
I see. This maybe too verbose for some people. I think we should have both - the 4-template arg variant and this one. This one can be added too later.

rolandschulz replied:
Why do we need a complicated `replicate` function with that many arguments at all? What's the advantage of this complicated function over just asking the user to write it in two steps>
```c++
s.select<W, S1>(i).replicate<R, S2>()
```
If reduced verbosity means it is harder to read it is counter-productive. We should optimize APIs for the reader of code.

kychendev replied:
Replicate is designed to enable programmer to specify arbitrary ISA region supported by Gen HW. Please refer to BSpec description: https://gfxspecs.intel.com/Predator/Home/Index/3010.  The suggested way of writing it in two steps is not equivalent.

kbobrovs replied:
In addition to @kychendev 's notes above

> Why do we need a complicated replicate function with that many arguments at all?

As I said, `replicate<2,2,2,2>` is much more compact and existing CM users find it quite handy. Plus this not the only place where region spec is used.

> If reduced verbosity means it is harder to read it is counter-productive.

this is quite subjective matter. Regions is a core widely used concept of the Gen SIMD programming. People will quickly get used to it, just be cause it will be used everywhere.

jasonsewall-intel replied:
This comes back to intention; is this extension to enable existing CM coders to put kernel source in the same TU as their DPC++ host code, or is this to expose Gen HW to users in general?   

If it is the latter, I think "people will get used to it" is not a great argument. I'd recommend that we expose something more verbose/understandable as @rolandschulz suggested and then provide a convenience wrapper to allow the 4-parameter lists that CM users are familiar with.

kbobrovs replied:
I added a TODO

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  template <typename __EltTy, int __Height, int __Width> auto format() &;
+
+  // Merge operations.
+  void merge(const value_type &__Val, const __mask_type_t<__N> &__Mask);
```
`__mask_type_t` isn't defined.

kbobrovs replied:
Right., because this is not the complete  source. The documentation for individual API elements will be built from source eventually, so there should be no such problems in future.

rolandschulz replied:
But it doesn't make sense to document an API which takes a type which isn't documented how it can be created. It's fine if the internals are not documented if they aren't needed for any of the documented APIs. But if the creating of the type isn't documented there is no documented way to use that API. You should either leave out that API for now or document at least the creation of that type.

kbobrovs replied:
> > `-fsycl-esimd` option option added to tweak FE behavior slightly compared to normal SYCL. In the future it is expected that no such distinction will be necessary, and the need in extra -fsycl-esimd option will go away.
> 
> I didn't see this discussed in the extension documentation. I think you should add some discussion of it there, and I'd recommend changing it to `-fsycl-explicit-simd` or just `-fsycl-simd`.

`-fsycl-explicit-simd` sounds better, as it accords with the feature name, will use that then

kbobrovs replied:
> But it doesn't make sense to document an API which takes a type which isn't documented how it can be created. 

@rolandschulz , I added missing definitions

Ruyk commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+    auto Acc1 = Buf1.get_access<cl::sycl::access::mode::read>(Cgh);
+    auto Acc2 = Buf2.get_access<cl::sycl::access::mode::read_write>(Cgh);
+
+    Cgh.single_task<class KernelID>([=] () EXPLICIT_SIMD {
```
Why a single_task with a macro at the end and not a different handler entry? `Cgh.explicit_simd`  ?

kbobrovs replied:
In our original proposal, there were new invocation APIs for explicit SIMD kernels - such as `parallel_for_task`, but after internal discussion it was decided to go with std SYCL APIs for invocation. From the technical standpoint, it is nice to have fewer APIs too.

Ruyk replied:
Well we could have a single "parallel_for" for everything but sometimes is better to distinguish entry points, specially if the programming model is very different. That macro in the lambda is easy to miss, specially if the kernel lambda is defined elsewhere.
Do you expect at some point that `single_task` without any attribute would be able to use this explicit simd style?
In the overall OneAPI / DPCPP / SYCL world, I am a bit concerned that people use this explicit SIMD thinking is just part of the standard, but it is easy to miss it is a (perfectly valid) vendor extension.

keryell replied:
Yet another typical use for kernel properties, for example based on https://www.youtube.com/watch?v=Fp8DuVWesT4 and https://gitlab.khronos.org/sycl/Specification/issues/296

I have an overdose of macros and user-visible attributes. :-)

kbobrovs replied:
The plan is to eventually have `h.single_task<...>([=]() [[intel::reqd_sub_group_size(1)]] {...})` define the explicit SIMD kernel w/o any extra compiler options, as SIMD kernel can be thought of as subgroup size 1 kernel. @rolandschulz actually suggested to replace EXPLICIT_SIMD with [[sycl_explicit_simd]] and then deprecate it once the above plan is supported.
So from that standpoint, standard SYCL invocation API seems better choice.

Ruyk replied:
This is a vendor extension, so clearly is your call. 
One of my concerns is that you could have the kernel separated from the single_task point of entry, and you may not even know what kind of kernel is that you are dispatching, just a SYCL kernel. Without a way for a C++ developer to identify the usage of this different API inside, they will be reduced to random runtime errors if the wrong kernel is used in the wrong device.

kbobrovs replied:
You are right, we already experience problems with distinguishing ESIMD kernels from others. Recently we discussed another approach internally, which is closer to parallel_for_task - use extra template parameter for parallel_for which will specify some "executor", which would be better aligned with modern trends.

kbobrovs replied:
added TODO to the bottom of the doc

Ruyk commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+Current ESP implementation does not support using certain standard SYCL features
+inside explicit SIMD kernels and functions. Most of them will be eventually
+dropped. What's not supported today:
+- Mixing `EXPLICIT_SIMD` kernels with SYCL kernels in a single source
```
so if you use this entry, you cannot use a normal parallel for in the same file? 
what about opencl interop?

kbobrovs replied:
Yes. This is a restriction of current implementation.

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  void merge(const value_type &__Val1, value_type __Val2,
+             const __mask_type_t<__N> &__Mask);
+} // namespace gen
+} // namespace cm
```
Wrong namespace.

kbobrovs replied:
thanks

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
````diff
@@ -0,0 +1,731 @@
+unsequenced with respect to the others.
+
+To reference a subset of the elements in simd vector object, Explicit SIMD provides ```select```
+function, which returns a simd or simd_view object (*described blow*) representing
````
```suggestion
function, which returns a `simd` or `simd_view` object (*described below*) representing
```

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Lots of weird formatting bugs in this paragraph (e.g. views)

kbobrovs replied:
thanks. copy-paste from PowerPoint

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Wrong namespace.

kbobrovs replied:
thanks

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Wrong namespace.

kbobrovs replied:
yes, will fix

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
We need to rename all the functions in this section (because they start with `cm_`).  I think we should use this as an opportunity to align with the reductions in `std::simd` and our other proposals.

Based on `std::simd` in [Parallelism TS v2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/n4808.pdf), we could use:
```
template <typename T1, int SZ, typename BinaryOperation>
reduce(simd<T1, SZ> v, BinaryOperation op);

reduce(simd<T1, SZ> v, std::plus<>()) // equivalent to cm_sum
reduce(simd<T1, SZ> v, std::multiplies<>()) // equivalent to cm_prod
reduce(simd<T1, SZ> v, sycl::maximum<>()) // equivalent to cm_reduced_max
reduce(simd<T1, SZ> v, sycl::minimum<>()) // equivalent to cm_reduced_min
```

We should limit support to these four built-in reduction operators initially, but this gives us a nice extension point for things like user-defined reduction operators.

If we think that having yet another overload of `reduce()` is unclear, we could call these something like `hreduce()` (horizontal reduce) or `simd_reduce()`.

kbobrovs replied:
@kychendev , please address

kbobrovs replied:
The  comments make sense to me.

Pennycook replied:
These comments haven't been addressed yet.  If @kychendev isn't opposed to the direction, I think we should rename these functions before the merge.  It seems like a small change to me.

kychendev replied:
I have updated the following reduction APIs as suggested:
reduce(simd<T1, SZ> v, std::plus<>())
reduce(simd<T1, SZ> v, std::multiplies<>())
For max/min I changed to the following as that's how it's defined in Parallelism TS v2, 
hmax(simd<T1, SZ> v)
hmin(simd<T1, SZ> v)

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Lots of formatting errors in this section again (e.g. T  element type).

kbobrovs replied:
thanks. pptx copy-paste problem

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
This comment shouldn't block this PR, but I'd like us to think more about what we can do to make these interfaces more user-friendly.

Two things stand out:
1) A user providing cache hints has to provide a lot more template arguments than required.  Could we make this nicer by providing the hints as tag-type arguments?
    ```c++
   // This works
   float* p;
   simd<uint32_t, 16> offsets;
   auto result = flat_load(p, offsets);

   // Adding cache hints makes it much more complex
   float* p;
   simd<uint32_t, 16> offsets;
   auto result = flat_load<uint32_t, 16, 1, CacheHint::Foo, CacheHint::Bar>(p, offsets);
    ```

2)  I find the existing names quite confusing.  I'd much prefer that the names were more obvious (e.g. `flat_load` could be called `gather` if that's all it does) and/or used overloading where appropriate to minimize the number of times a reader of code has to check documentation to figure out what a function is supposed to be doing.

kbobrovs replied:
Makes sense to me. @kychendev, please comment.

kychendev replied:
I'm open to change the name to make it more user friendly. The cache hints setting are mostly based on existing implementation. Please check with Gang who added these memory access APIs.

cmc-rep replied:
I do not see why gather/scatter is better naming than load/store. Our current naming scheme has its own consistency.  I also do not see why tagged argument is any better than template

Pennycook replied:
The more explicit and obvious a name is, the better.  Anybody reading a function called "gather" or "gather load" knows that the function being called performed is a gather operation.  I wouldn't know that a "flat load" was a gather without reading the documentation.

A tagged argument is better in this case because it allows for more template arguments to be deduced.  Expanding on my previous example:

```c++
// Without cache hints, type and length can be deduced from offsets
float* p;
simd<uint32_t, 16> offsets;
auto result = flat_load(p, offsets);

// With cache hints as templates, verbosity increases significantly:
// - Providing any cache hint forces the user to specify the type and length
float* p;
simd<uint32_t, 16> offsets;
auto result = flat_load<uint32_t, 16, 1, CacheHint::Foo, CacheHint::Bar>(p, offsets);

// With cache hints as tag types, verbosity is reduced:
// - Providing a cache hint does not prevent deduction of type and length
float* p;
simd <uint32_t, 16> offsets;
auto result = flat_load(p, offsets, CacheHint::Foo{});
```

Note also that the templated form prevents a developer from specifying an L3 hint without also explicitly specifying an L1 hint.  If `flat_load` accepted a list of hints, it might be possible to refactor the hints to specify them in any order, and it may be more extensible to future cache hints:
```c++
auto result = flat_load(p, offsets, l3_cache_hint_foo, ...);
```

cmc-rep replied:
load/store means gather/scatter unless it is block_load/block_store

Pennycook replied:
I understand that, and the distinction is clear to somebody who has read the documentation.

I'm arguing that ideally we wouldn't be forcing developers to read documentation in order to make sense of what code does.

cmc-rep replied:
Yeah, who can write code with reading documents? especially with modern C++.

Pennycook replied:
I didn't say _write_ code -- I said _read_ code.  Code is read more than it is written, and having good names for functions helps people who didn't write a particular piece of code to understand what it's supposed to do.  Having good names is helpful for training material, code review, onboarding new programmers, etc.

I think these names could be better, and I think we'll waste time explaining that `flat_load` is a gather operation every time we show somebody an example code using explicit SIMD types.  This isn't a hill I'm prepared to die on, so I'll leave it at that.

jasonsewall-intel replied:
@cmc-rep are you being facetious about reading documentation?  Descriptive names and interfaces are very important for readable/maintainable/accessible code. 

'flat' addresses are only meaningful on Gen, and the convention that load == gather is obviously a Gen-specific notion too, since we keep having to say 'oh this is a gather' over and over just in this conversation! 

Maybe it would be best to change the name of this extension back to DPC++-CM?

keryell replied:
Reading documentation is very dangerous when designing/writing write-only programs... :-)

rolandschulz replied:
Also another area where it needless differs from `std::simd`. Non-gather load/store is called `copy_from` and `copy_to`. Why invent new names just to confuse people?

keryell replied:
> ```c++
> // Without cache hints, type and length can be deduced from offsets
> float* p;
> simd<uint32_t, 16> offsets;
> auto result = flat_load(p, offsets);
> auto result = flat_load<uint32_t, 16, 1, CacheHint::Foo, CacheHint::Bar>(p, offsets);
>```

An approach *à la* https://github.com/chriskohlhoff/propria from @chriskohlhoff would be to add a property to the pointer, such as

```c++
auto result = flat_load(p, offsets);
auto result = flat_load(decorate<CacheHint::Foo, CacheHint::Bar>(p), offsets);
```
The advantage is that you do not have to change all tour API and all the uses of this
decorated pointer will benefit from this.
`decorate` is to be bikeshed accordingly.

kbobrovs replied:
added TODO to the headers

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
We should look into what can be done to simplify these atomic functions and align their design with the other new atomic features.  That is perhaps out of scope for this PR (the direction is less clear than for the `reduce` changes, for example) but we should open an issue to track it.

kbobrovs replied:
@Pennycook , can you please open the issue then?

Pennycook replied:
I'll open it as soon as this is merged.

kbobrovs replied:
thanks

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```suggestion
Explicit SIMD extension supports "private global" variables - file scope
```

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Why is this not included automatically when the extension is enabled?

kbobrovs replied:
Following mainstream SYCL - sycl.hpp is not auto included under -fsycl. I can add a TODO and file an issue.

Pennycook replied:
That's a good argument, but we currently do automatically include things like the sub-group and USM headers.  I don't mind which way we go as long as we're consistent; marking this with a TODO and filing an issue to address this later works for me.

kbobrovs replied:
Filed internal (for now) issue, added a TODO

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```suggestion
- Design interoperability with SPMD context
```

Same reason for SIMT => SPMD change as before.  There isn't a public proposal for this functionality yet and so I don't think we should give it a name (yet).

Pennycook commented:
>   `-fsycl-esimd` option option added to tweak FE behavior slightly compared to normal SYCL. In the future it is expected that no such distinction will be necessary, and the need in extra -fsycl-esimd option will go away.

I didn't see this discussed in the extension documentation.  I think you should add some discussion of it there, and I'd recommend changing it to `-fsycl-explicit-simd` or just `-fsycl-simd`.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  further widening by the compiler, as with traditional SIMT programming.
+- Low-level APIs efficiently mapped to Gen architecture, such as block reads.
+
+One of the future directions is enabling this extension for other architectures,
```
ok. But I'd rather leave out details such as where for now.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+other devices will result in error. 
+
+Kernels and `SYCL_EXTERNAL` functions using ESP must be explicitly marked with
+the `EXPLICIT_SIMD` attribute macro. Subgroup size quiery within such
```
To have flexibility in implementation. At this point (and likely in mid-term too) we need to distinguish SYCL `reqd_sub_group_size(1)` kernels from ESIMD kernels. So `EXPLICIT_SIMD` expands to `sycl_explicit_simd` attribute and then FE adds `reqd_sub_group_size(1)` to all kernels marked with `sycl_explicit_simd`.

rolandschulz replied:
Why not ask people to put the `sycl_explicit_simd` attribute for now? When it isn't needed anymore we can make it a deprecated alias for `reqd_sub_group_size(1)`. The marco on the hand can't be deprecated. And in general we try to avoid macros.

kbobrovs replied:
With macro we can just redefine EXPLICIT_SIMD to `reqd_sub_group_size(1)` later w/o any changes to the user code.

rolandschulz replied:
Yes but then `EXPLICIT_SIMD` stays in the user code rather than gets a deprecation warning. That's a disadvantage not an advantage.

kbobrovs replied:
OK, I will make the change.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+- Constant accessors
+- `sycl::accessor::get_pointer()`. All memory accesses through an accessor are
+done via explicit APIs; e.g. `sycl::intel::gpu::block_store(acc, offset)`
+- Few others (to be documented)
```
No. This will be added incrementally.

rolandschulz replied:
Are they unknown or why would we not document all of them? If all known ones are documented that's fine.

kbobrovs replied:
Some potential ones are known - need to be checked before specifying. I admit there can be some unknowns. 
This is WIP rather than a complete formal specification, will be improved/extended over time.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
````diff
@@ -0,0 +1,731 @@
+
+### SIMD vector class
+
+The ```sycl::intel::gpu::simd``` class is defined as a templated class for a vector
+of fundamental data elements. The fundamental data element type **__Ty** must be
+vectorizable, which can be any of the C++ character/integer/real floating-point types,
+or IEEE-754 half-precision floating-point type. The length of the vector
+is a constant expression, as determined by the template parameter **__N** in a
+given specialization.
````
@kychendev, please address these comments

kychendev replied:
OK, will update this as suggested.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
````diff
@@ -0,0 +1,731 @@
+files (GRF).
+
+```cpp
+namespace cm {
````
ok

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+
+Explicit SIMD execution model is a basically an equivalent of the base SYCL
+execution model with subgroup size restricted to 1. Which means in ESP
+it is guaranteed that there is a single work item in any workgroup, and it maps
```
You are right, this is just a typo, obviously.

Pennycook replied:
Ok, thanks for the clarification.  I genuinely wasn't sure what the restrictions of the initial implementation were!

It might be useful to add a simple example showing the mapping between an ND-range and the number of thread-groups and EU threads, and showing the usage of explicit SIMD together with work-group barriers and SLM.  That would help to highlight what's possible with this extension.

kbobrovs replied:
ok, will try. Will add a TODO for now.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+it is guaranteed that there is a single work item in any workgroup, and it maps
+to a single hardware thread. All standard SYCL APIs continue to work,
+including `sycl::intel::sub_group` ones, which become either a no-op or
+trivial. E.g. a barrier becomes just a memory fence for a compiler, collectives
```
work-group barrier will behave as you expect

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,745 @@
+};
+
+template <int __N>
+using __mask_type_t = typename __vector_type<uint16_t, __N>::type;
```
The mask should also be a wrapper around the clang-vector type rather than the clang-vector type itself.

The internal storage should be implementation defined. `uint16_t` is a bad choice for some HW. Nor is it how clang-vector types works (using the same size int as the corresponding vector type used for comparison (e.g. long for double and int for float)).

kbobrovs replied:
I added a TODO

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+public:
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
```
What's the reason to wrap the clang vector type twice? Why not store the clang vector type directly in the SIMD type?

kychendev replied:
It's defined in other internal file. We also need to perform additional type property checks in the wrapper struct for supported vector types.

jasonsewall-intel commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Can this be parameterized? Why is 4 magic?

kbobrovs replied:
4 maps to Gen hardware memory access instruction (message)

jasonsewall-intel replied:
I understand, but is that what it is going to be like for all future hardware? Would it be wise to parameterize this so that we don't need new function names for flat_loadX at some point?

kbobrovs replied:
I'm not sure. For there is 1 and 4 only, and people would have to use more verbose templated syntax anticipating other widths which is not clear if/when will happen.

jasonsewall-intel replied:
Can't this be deduced?

kbobrovs replied:
Probably. Do you have an example?

jasonsewall-intel commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
This looks like a find+replace error.

kbobrovs replied:
thanks

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```suggestion
#include <sycl_esimd.hpp>
```

kychendev commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
````diff
@@ -0,0 +1,731 @@
+```cpp
+namespace cm {
+namespace gen {
+template <typename __Ty, int __N> class simd {
````
Will remove leading underscores to conform to the naming convention.

kychendev commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &__Val);
+  constexpr simd(std::initializer_list<__Ty> __Ilist) noexcept;
```
It's an implementation choice so we can leverage std initializer list syntax for simd vector constructor (e.g., simd<int, 16> v3{0, 2, 4, 6, 1, 3, 5, 7};). These constructors have a simple implementation which are expected to be fully optimized by compiler without throwing.

rolandschulz replied:
Providing both an overload of initializer-list and the same type itself causes really weird behavior.
E.g.

    simd s1(1,2); //calls next constructor
    simd s2{1,2}; //calls this constructor
This might not be confusing for all users but to everyone using uniform-initialization syntax. Therefore if you want to use this constructor the other one should have a special type (see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es64-use-the-tenotation-for-construction) to avoid this issue.

Also this seems like one of those areas where this simd-type needless differs from std::simd. Why should these constructors be different? Why reinvent the wheel and have all the work of fixing these problems if we could just use the existing solution. Especially if that is anyhow the long-term goal. Adding extra stuff like the select is totally fine. But differ on things which have no apparent advantage and aren't as thought through seems to have only downsides.

> These constructors have a simple implementation which are expected to be fully optimized by compiler without throwing.
And the other constructors (like the copy-constructor) which aren't declared as noexcept, they might throw?

kbobrovs commented:
@rolandschulz , please take a look at the changes @kychendev pushed recently to address your comments

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,743 @@
+Here are some future directions in which this API is intended to evolve:
+- enabling this extension for other architectures, such as x86, with extracting
+  and clearly marking generic and target-dependent API portions
+- aligning with `std::simd` and maybe providing `std::simd` implemenation atop
```
```suggestion
- aligning with `std::simd` and maybe providing `std::simd` implementation atop
```

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,731 @@
+
+## Explicit SIMD execution model
+
+Explicit SIMD execution model is a basically an equivalent of the base SYCL
```
Was this suggestion applied?  I still see the typo, here.

kbobrovs replied:
Probably that one was missed - thanks.

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,743 @@
+other devices will result in error. 
+
+Kernels and `SYCL_EXTERNAL` functions using ESP must be explicitly marked with
+the `[[sycl_explicit_simd]]` attribute. Subgroup size query within such
```
I don't want to bikeshed on this, but just a note: I think `[[sycl_explicit_simd]]` is better than a macro (for all the reasons @rolandschulz pointed out) but the spelling here is different to attributes like `[[intel::reqd_work_group_size]]`.

We should decide what rules we're following for attributes (if any).  I'll open an issue for this as well.

bader replied:
I think we should also update all the examples to align with this proposal: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/KernelRHSAttributes/SYCL_INTEL_attribute_style.asciidoc

kbobrovs replied:
Done. (@bader - I assume alignment means function type attribute placed before definition instead of function attribute placed before the declaration)

Pennycook commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```suggestion
- Consider auto-inclusion of sycl_explicit_simd.hpp under -fsycl-esimd option
```

kbobrovs commented:
> > `-fsycl-esimd` option option added to tweak FE behavior slightly compared to normal SYCL. In the future it is expected that no such distinction will be necessary, and the need in extra -fsycl-esimd option will go away.
> 
> I didn't see this discussed in the extension documentation. I think you should add some discussion of it there, and I'd recommend changing it to `-fsycl-explicit-simd` or just `-fsycl-simd`.

This one is implemented in https://github.com/intel/llvm/pull/1743

kbobrovs commented:
@keryell, @Ruyk, there is only documentation for prior incarnation called "C-for-Metal". Some links:
https://software.intel.com/content/www/us/en/develop/articles/intel-c-for-metal-the-precious-metal-for-computing-on-intel-graphics-cards.html

https://01.org/c-for-metal-development-package/downloads/cm-programming-reference-1

 @kychendev, please add if you are aware of anything else

kbobrovs commented:
> These comments haven't been addressed yet. If @kychendev isn't opposed to the direction, I think we should rename these functions before the merge. It seems like a small change to me.

@Pennycook, I think @kychendev updated the reduction APIs - please take a look

Pennycook commented:
> @Pennycook, I think @kychendev updated the reduction APIs - please take a look

Looks great, thanks!

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Is this intentional not a correct copy constructor (would need to be const for that)? 

I believe we agreed that `simd_view` would have a deleted copy and move constructor. Why are they suddenly back?

kbobrovs replied:
added TODO to the headers

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Why is this and the next constructor public? Those should only be called internally by e.g. `select`, correct?

kbobrovs replied:
added TODO to the headers

rolandschulz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
same as above. Why does the non-const overload not return a simd_view to make it writeable?

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,776 @@
+Here are some future directions in which this API is intended to evolve:
+- enabling this extension for other architectures, such as x86, with extracting
+  and clearly marking generic and target-dependent API portions
+- aligning with `std::simd` and maybe providing `std::simd` implementation atop
```
to avoid re-implementing lots of stuff and take advantage of sycl::intel::gpu::simd efficiency

rolandschulz replied:
Implementing it on top of `gpu::simd` is the strategy which requires a reimplementation of `std::simd` features missing from `gpu::simd`. On the other hand there is an existing implementation directly on top of gcc/clang vector: https://github.com/VcDevel/std-simd . We should explore how we can reuse most of that implementation also for the GPU.

mattkretz replied:
This is a pretty fundamental design question, I'm sorry if I am missing context, but I just found out your work today. First of all, :+1:  thanks for working on a SIMD type for GPU. I wanted to do this since years but didn't have the manpower/time.

What I'd like to know is why you can't (don't want to) take the `std::experimental::simd` specification as is. I'd be very happy to help. Especially with folding back improvements and extensions into C++ committee papers targeting C++23. If the `simd<T, ?>` type is 1:1 portable between CPU and GPU that would be a huge win in terms of software maintenance, and freedom to move to newer platforms.

kbobrovs replied:
The short answer is that the primary target for this specification is Intel GPU, which is reflected in the namespace.

But generalizing the spec (and implementation) to fit more devices and be more compatible
with std C++ is definitely welcome, as long as it does not hurt GPU performance. simd<T, ?>
type is portable to CPU, but performance of various associated APIs is another question. Our
plan is to upstream the existing initial implementation, then iteratively improve it along with
the specification, making sure it remains functional and performant on Intel GPU.

mattkretz replied:
> The short answer is that the primary target for this specification is Intel GPU, which is reflected in the namespace.

Right. I think it totally makes sense to explore what would be the perfect API for programming a specific target. It makes less sense to write applications with it. Because portability and being able to move to newer/other hardware is an important reason for using SIMD types in the first place.
Also, "Intel GPU" is a narrow target right now. But do you know whether it will stay that way?

> But generalizing the spec (and implementation) to fit more devices and be more compatible
> with std C++ is definitely welcome, as long as it does not hurt GPU performance.

:+1: A SIMD type that doesn't allow you to reach full performance is a failure (or hopefully just buggy and needs to be fixed). Performance is the reason why we use it.

I guess what I'd like to discuss is: *If you deviate from the Parallelism TS 2, why? Is it to simplify your implementation, is it because of missing functionality, is it because of performance, or?* This is important feedback to the C++ committee when merging the TS into the IS is discussed.

cmc-rep replied:
To help you understand our points of view, I recommend you take a look the documents and examples at https://github.com/intel/cm-compiler

keryell replied:
Be sure that all the interesting links which were mentioned in the discussions end up in the extension description itself, so it is not lost for the future.

kbobrovs replied:
Added these comments to the bottom of the doc as TODO

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,776 @@
+
+The `sycl::intel::gpu::simd` class is a vector templated on some element type. The element type must be vectorizable type. The set of vectorizable types is the set of fundamental SYCL arithmetic types (C++ arithmetic types or `half` type) excluding `bool`. The length of the vector is the second template parameter.
+
+Each simd class object is mapped to a consecutive block of general register
```
reworded

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
````diff
@@ -0,0 +1,776 @@
+</p>
+
+```cpp
+a.select<4, 2>(0) = b;  // selected elements of a are replaced
````
The naming sounds strange for something that can be in LHS. On the other hand I do not have another good proposition...

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
It sounds like a specialization of the `select` above.
And one can ask how it relates to `std::format`...

iburyl replied:
It is what C++20 calls `std::bit_cast`, not the select. Select takes subset of values, while format - bit_casts them (into potentially differently sized set) and reshapes them (but in overloads with more template arguments).

By the way, format is a function - where are parentheses?

Ruyk replied:
Why not call it bit_cast then?

kbobrovs replied:
added TODO to the headers

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
This could be `assign` with a mask.

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
This is where I was expecting... `select`! :-)

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
How is it different from your `select` above?
I prefer this `simd_view` naming.

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Inconsistent naming style with SYCL.

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
The compiler

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
What would be `vmax` then? Does it change if we turn the processor? :-)
Would `inner_max` or something like that would make sense?
I do not know whether @mattkretz looked at these kind of operations and studied the naming.

mattkretz replied:
The names `hmax` and `hmin` were discussed at some point in the Parallelism TS 2 process. I don't remember exactly, but I'm pretty sure there wasn't much concern. I agree the names are not perfect. (sounds like useful [TS feedback](https://github.com/mattkretz/std-simd-feedback/issues/16) :wink: ) `inner_min/max` sounds too mathy IMHO but still a bit more obvious. `reduce_max` or `max_reduce` are also not obviously great. (`valarray` has member functions `min()` and `max()`. But member functions inhibit writing generic code that also works with arithmetic types.) `min/max_element(simd)` maybe? In any case, would be great to align with what (hopefully) goes into C++23.

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
What do you mean by "file scope" in the context of SYCL?

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Would you mind a
```c++
intel::gpu_private<32, simd<int, 16>> vc;
```
to be bikeshed instead of this C++ alien syntax?

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```c++
constexpr auto Size = 128U;
```
and so on

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Please use the terse syntax for this if you want to promote USM :-)

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```suggestion
  for (auo i = 0; i != Size; i++) {
```
so we are safe even for `Size = 0`

kbobrovs replied:
done

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Is it not possible to infer `float` and `VL`?

kbobrovs replied:
Compiler complains that it can't determine vector element type

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
3 `auto` in LHS? I guess you want to sell your extension... :-)

kbobrovs replied:
changed to explicit type

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Idem

kbobrovs replied:
done

keryell commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
Do you really need the statistics above as an example for a specification?

kbobrovs replied:
removed

keryell commented:
Definitely useful for SIMD architectures and... FPGA! :-) @aisoard @javier-cabezas
It would be nice to converge on the naming at some point.

mattkretz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,776 @@
+  simd &operator |=(const simd &RHS);
+  simd &operator ^=(const simd &RHS);
+
+  // Compare operators.
+  simd<uint16_t, N> operator >(const simd &RHS) const;
+  simd<uint16_t, N> operator >=(const simd &RHS) const;
+  simd<uint16_t, N> operator <(const simd &RHS) const;
+  simd<uint16_t, N> operator <=(const simd &RHS) const;
+  simd<uint16_t, N> operator ==(const simd &RHS) const;
+  simd<uint16_t, N> operator !=(const simd &RHS) const;
```
Returning `simd<uint16_t, N>` looks rather target-specific to me. Are you certain this needs no further abstraction to make it more portable?

iburyl replied:
At least it should be `mask_type_t`, which might become more abstracted in the future revisions.

kbobrovs replied:
added TODO

mattkretz commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
```diff
@@ -0,0 +1,776 @@
+  simd operator--(int);
+
+  // Binary operators.
+  auto operator +(const simd &RHS) const;
```
ditto for compares

kbobrovs replied:
added TODO to the headers

kbobrovs commented:
@Ruyk , @keryell , @mattkretz , @iburyl , @rolandschulz , and other reviewers - apologies for delay.
All review comments are greatly appreciated, thank you all!
Some of them have been addressed in the PRs with actual SIMD library code, and will be propagated to the spec shortly:
https://github.com/intel/llvm/pull/1853
https://github.com/intel/llvm/pull/1854
https://github.com/intel/llvm/pull/1855
https://github.com/intel/llvm/pull/1856
(At some point we should create automated API spec generation from sources to avoid duplicating in two places.)

This is initial experimental version to try out functionality and performance. We are going to work with the community on incrementally improving the APIs to bring them closer to std C++ and SYCL in the next several months. At this point we would like to
- "freeze" the experimental APIs until
   * we have some working version in this branch
   * we have initial performance regression suite in place to check API changes
- and
   * __collect all unresolved review comments and add them to the TODO section in this document__
   * merge the current state of the experimental spec

Then review can be resumed and we can then keep adding comments to this document and/or the sources.

jbrodman commented:
Will approve after @rolandschulz does.

kbobrovs commented on sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md:
````diff
@@ -0,0 +1,776 @@
+
+*Functor kernel*
+```cpp
+using AccTy = cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
````
fixed

kbobrovs commented:
@rolandschulz, @Pennycook, @keryell, @Ruyk, @iburyl, @mattkretz, @jasonsewall-intel - sorry for delay. I updated the spec, please take another look. Specifically:
- removed the API definition, as we now have sources. The spec will just reference the auto-generated API docs
- addressed some of the comments
- summarized others in the end of the doc
- moved your comments to specific APIs as TODOs to the code, please see the following patches:
   * https://github.com/intel/llvm/pull/1853/commits/3f263a4eeffc90999c9c308f7b9730c9b5eb965a
   * https://github.com/intel/llvm/pull/1854/commits/92180f46b0862e474c9540688751e40d53a8c13e
   * https://github.com/intel/llvm/pull/1856/commits/2f1b56cadafc1710f83e1d43ec330e0a48d0bf15
